compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs

```diff
@@ -1601,77 +1596,150 @@ impl EnumMemberDescriptionFactory<'ll, 'tcx> {
                 ref variants,
                 tag_field,
             } => {
+                let calculate_niche_value = |i: VariantIdx| {
+                    if i == dataful_variant {
+                        None
+                    } else {
+                        let value = (i.as_u32() as u128)
+                            .wrapping_sub(niche_variants.start().as_u32() as u128)
+                            .wrapping_add(niche_start);
+                        let value = tag.value.size(cx).truncate(value);
+                        // NOTE(eddyb) do *NOT* remove this assert, until
+                        // we pass the full 128-bit value to LLVM, otherwise
+                        // truncation will be silent and remain undetected.
+                        assert_eq!(value as u64 as u128, value);
+                        Some(value as u64)
+                    }
+                };
+
+                // For MSVC, we will generate a union of two structs, one for the dataful variant and one that just points to
+                // the discriminant field. We also create an enum that contains tag values for the non-dataful variants and
+                // make the discriminant field that type. We then use natvis to render the enum type correctly in Windbg/VS.
```

Comment:
For the benefit of debugger authors, would be nice to document this in terms of C structs/unions.   Also, please spell out the rules of how to decode them.

I am guessing the above means something like:
```
union <enum name> {
    struct <variant name N> {
         <field 1>
         <field 2>
         ...
    },
    struct discriminant$ {
        enum tag$ {   
            <variant name 1>
            <variant name 2>
            ...
        } discriminant;
    }
}
```
?